### ISSUE TITLE
Search in a few first lines only

### ISSUE THREAD

**Issue by abitrolly** [ggreer/the_silver_searcher#1337]
I am trying to find all SQLite dababases on my disk with `ag` with the following command.
```
$ ag "SQLite format 3" / --silent --search-binary --files-with-matches --literal --ignore "*.so"
```
It is too slow, because it scans the whole file contents. Is it possible to limit the search to the first few lines or bytes? I couldn't find the option.

**Comment by genivia-inc** [comment]
You could use `-m NUM` to skip the rest of the file after NUM matches. I know that this isn't exactly what you are asking for, but it may help when you have a hunch how many matches to expect up to a certain number of lines in the input file.

I suppose [ugrep option -K](https://github.com/Genivia/ugrep#max) is what you are looking for, which doesn't yet exist in ag.

**Comment by abitrolly** [comment]
`ugrep -K` helps to search in the first few lines.
```
-K NUM1[,NUM2], --range=NUM1[,NUM2]
        Start searching at line NUM1 and end at line NUM2 when specified.
```
For searching in the first 2048 bytes maybe `ag` allows to filter contents of each matched filename through `head -c 2047`?

**Comment by genivia-inc** [comment]
> For searching in the first 2048 bytes maybe ag allows to filter contents of each matched filename through head -c 2047?

Nope. I don't believe `ag` offers a built-in filter mechanism to run the input through a filter utility when searching files with recursive searches. However, you can pipe the output of `head` to `ag` one file at a time. For recursive searches you will need to write a shell script to recursively find the files to search first, e.g. with `find`, then loop over them to execute `head -c 1047 | ag ...`.

Some other greps offer filters, such as [ugrep option --filter](https://github.com/Genivia/ugrep#filter) and [ripgrep option --pre](https://github.com/BurntSushi/ripgrep) (see ripgrep `--help`). For example, with ugrep you should be able to do a `ugrep -rl -F "SQLite format 3" --exclude='*.so' --filter='*:head -c 2047'` and get the results you're looking for.
